readme.md
# Manual de zPortapapeles

Este complemento tiene la base del complemento Fake Clipboard Announcement.

Con zPortapapeles tendremos el anuncio de las pulsaciones de copiar, pegar, deshacer, cortar y seleccionar todo.

Se incluyo al complemento la posibilidad de activar y desactivar sonidos que refuercen los mensajes al igual que un historial con el que podremos copiar al foco.

## Gestos de entrada...

En el apartado de Gestos de entrada... de NVDA si buscamos la categoría zPortapapeles podremos añadir una combinación de teclas que no viene asignada por defecto al historial para mostrar el dialogo.

También tendremos un apartado donde poder modificar las teclas disparadoras referidas al portapapeles, este apartado solo modificarlo si nuestro lenguaje o sistema tiene otras teclas asignadas por defecto al portapapeles.

## Opciones del complemento

En las opciones del NVDA en la categoría Opciones de zPortapapeles podremos activar y desactivar a través de casillas de verificación tanto si deseamos sonidos como si deseamos tener el historial activado, también se podrá elegir si se desean recibir mensajes hablados.

Si la casilla del historial esta activada tendremos un cuadro combinado en el cual elegir el tiempo de monitoreo del portapapeles y una casilla para activar o desactivar si queremos un sonido cuando se agregue algo al historial.

Si la casilla esta desactivada tanto el cuadro combinado como el poder elegir si queremos sonidos en el historial no aparecerán.

Advertir que si empezamos a recibir errores del portapapeles es conveniente aumentar el tiempo de monitoreo del portapapeles.

## Dialogo del Historial

Este dialogo tendremos que asignarle en gestos de entrada una combinación de teclas por defecto no viene asignada.

El dialogo solo se abrirá cuando tenga una entrada guardada, mientras no tenga entradas no podremos abrirlo.

Todo se nos informara con mensajes informativos al igual si ya tenemos el dialogo abierto.

Una vez abierto el dialogo consta de la lista de entradas que tenga y 4 botones que son los siguientes:

* Borrar o Alt+B: Si pulsamos este botón borrara la entrada que la lista tenga el foco.

* Borrar todo o Alt+T: Borrara todas las entradas del historial.

* Refrescar o Alt+R: Refrescara el historial con las nuevas entradas agregadas, esto va bien por si agregamos al historial mientras el dialogo esta abierto.

* Cerrar o Alt+C, Escape o Alt+F4: Cerrara el dialogo del historial.

Cuando estamos en la lista si pulsamos INTRO la entrada que tenga el foco se copiara a la aplicación que este detrás del dialogo de historial.

Por ejemplo si tenemos abierto el bloc de notas y abrimos el historial y pulsamos INTRO en la primera entrada copiara al bloc de notas el item que tengamos seleccionado.

## Limitaciones de zPortapapeles

zPortapapeles no es compatible con otros complementos que estén instalados y que hagan lo mismo, como Fake Clipboard Announcement o Clipspeak. Puede haber otros, y en este caso tendrá que desactivarlos si desea usar zPortapapeles.

Apuntar que las entradas en el historial se borrarán cuando reiniciemos NVDA.

El historial de portapapeles de Windows puede tener entradas duplicadas si usamos el historial de zPortapapeles. Tendremos que elegir cual deseamos usar.

## Agradecimientos

* Javi Domínguez: Por su ayuda desinteresada y por contribuir con la función de monitoreo del portapapeles.
* Portugués Brasil: pedro-hdias
* Ruso: Valentin Kupriyanov
* Turco: Umut KORKMAZ
* Italiano: Alessio Lenzi
* Francés: Rémy Ruiz
* Árabe: Wafiq Taher

# Registro de cambios.
## Versión 0.5.1.

* Actualizado idioma turco.

## Versión 0.5.

* Agregadas  estadísticas del texto seleccionado

Para usar esta característica tendremos que asignar un gesto en el Gestor de entradas de NVDA en la categoría zPortapapeles y Estadísticas del texto seleccionado.

Cuando tengamos un texto seleccionado si pulsamos la combinación que hayamos asignado nos informara de:

- Caracteres sin espacios de la selección.
- Caracteres con espacios de la selección.
- Número de palabras.
- Número de líneas.
- Número de párrafos.

* Agregada la posibilidad de copiar partes de un texto seleccionado.

Para poder usar esta función tendremos que agregarle un gesto en el gestor de entradas.

Tendremos que ir a la categoría zPortapapeles y agregar el gesto que deseemos a la opción Ejecuta el dialogo para copiar al portapapeles parte de un texto seleccionado.

Bien cuando tengamos un texto seleccionado y lancemos esta opción se abrirá una ventana con una lista que contendrá todas las palabras del texto seleccionado.

Dicha lista tiene casillas de selección por lo que tendremos que seleccionar aquellas palabras que deseemos para copiar al portapapeles.

El texto que recibe este dialogo esta limpio de símbolos de puntuación por lo que tendremos que tener cuidado si alguna palabra viniera con puntuación por ejemplo esto hay que tener cuidado si seleccionamos un texto de programación y viene algo así:

		if event.GetUnicodeKey()

event.GetUnicodeKey() se mostrara en la lista así:

eventGetUnicodeKey habiendo quitado el punto y los paréntesis.

Esta opción principalmente esta enfocada a texto simple, pero si se solicita puedo agregar una opción que muestre el texto con símbolos de puntuación.

El dialogo no podrá ser abierto si ya tenemos un dialogo de historial abierto y viceversa.

El texto que se copia al portapapeles es en orden de la selección original indiferentemente en que orden hayamos elegido los elementos en el dialogo.

Por ejemplo, la frase:

Hola, como estás, el perro tiene hambre

Si deseamos seleccionar Hola, perro y hambre da igual que seleccionemos primero, si marcamos las tres se copiará al portapapeles indiferentemente del orden que lo hayamos seleccionado, siendo lo siguiente lo que se copiará al portapapeles:

Hola perro hambre

Este dialogo necesita por lo menos una selección de dos palabras para ser llamado.

Una vez en el dialogo es necesario al menos tener una palabra seleccionada para copiar al portapapeles.

* Añadida compatibilidad con NVDA 2023.1

## Versión 0.4.2.

* Agregado la posibilidad de leer el contenido de portapapeles incluido portapapeles de más de 1024 caracteres.

## Versión 0.4.1.

* Solucionado problema al guardar el tiempo de refresco del modo juego
## Versión 0.4.

* Agregado modo juego.

Para activar dicho modo tendremos que asignarle un gesto en el gestor de entradas.

Este modo ira bien para aquellos juegos que copian al portapapeles y necesitan de traducción.

En opciones del complemento tendremos la posibilidad de aumentar el refresco de actualización del portapapeles y la posibilidad de elegir el idioma destino de la traducción.

El idioma origen en este caso del juego se detectara automáticamente y se traducirá lo copiado al portapapeles al idioma que hayamos elegido.

Mientras el modo juego esta activado todas las ordenes del resto de complemento quedan deshabilitadas incluido el historial del portapapeles.

## Versión 0.3.1.

* Se reescribió la función de impedir la ejecución del complemento en pantallas seguras.

* Agregado idioma Árabe.

## Versión 0.3.

* Agregado detección si hay texto seleccionado al copiar.

Ahora el complemento cuando demos a copiar si no detecta texto nos informara con un mensaje.

* Agregado el poder desactivar y activar rápidamente las funciones del complemento respecto al portapapeles.

En gestos de entrada podremos configurar una combinación para activar y desactivar rápidamente lo referido al portapapeles.

Cuando lo desactivemos el portapapeles y los mensajes serán los nativos de Windows junto a los mensajes que NVDA tenga predefinidos para el portapapeles.

Esto no afectara al historial que si esta activo seguirá activo.

* Agregado compatibilidad con teclados que usen caracteres cirílicos.

Se soluciono un problema en distribuciones de teclado que usan caracteres cirílicos.

* Agregados idiomas Ruso, Turco, Italiano y Portugués Brasil.

## Versión 0.2.

* Cambiadas todas las funciones del portapapeles a cTypes

Ahora el portapapeles se manejara directamente con funciones del sistema evitando usar las funciones de NVDA y wxpython.

* Preparado el complemento para ser traducido.

## Versión 0.1.5.

* Hablar lo copiado al portapapeles y agregado al Historial.

Si tenemos activada la opción Activar o desactivar el historial del portapapeles en las opciones del complemento ahora tendremos esta nueva opción.

Con esta opción lo copiado al historial o si ya se encuentra en el historial lo ultimo copiado al portapapeles se leerá con NVDA.

Solo se mencionara lo ultimo copiado y una única vez para no saturar hasta que no vuelva a decir otra cadena copiada no se volverá a repetir la anterior.

* Se borrara el portapapeles cuando inicie el complemento por primera vez.

Ahora ya no se copiara lo que tuviéramos en el portapapeles cuando iniciábamos NVDA, ahora el portapapeles empezara limpio.

Esto no afecta al historial de portapapeles de Windows.

*** Advertencia: esto tenemos que tenerlo en cuenta ya que se puede borrar algo importante que tuviéramos en el portapapeles. ***

## Versión 0.1.4.

* Cambiada la forma de obtener los datos del portapapeles.

## Versión 0.1.3.

* Se subió el requisito del complemento a NVDA 2021.2 como versión mínima para poder ser usado.

* Se agregaron opciones para el historial.

Ahora podremos elegir el tiempo de monitoreo como si deseamos un sonido cuando algo se copie al historial.

* Se agrego compatibilidad con Word y Excel para que no anuncie las teclas del portapapeles.

* Ahora se monitorea el portapapeles por lo que cualquier complemento que copie al portapapeles ya es compatible, incluso el copiado de NVDA al foco

## Versión 0.1.2.

* Se agrego Activar o desactivar anuncios hablados del portapapeles en opciones

En las aplicaciones permitidas nos omitirá los mensajes correspondientes a las teclas del portapapeles.

## Versión 0.1.1.

* Se cambio la forma de gestionar las capturas del portapapeles y su gestión.

* Se arreglo el error que dejaba abierto el portapapeles.

## Versión 0.1.

* Versión inicial.

changelog.md
* Actualizado idioma Turco.

buildVars.py
# -*- coding: UTF-8 -*-

# Build customizations
# Change this file instead of sconstruct or manifest files, whenever possible.


# Since some strings in `addon_info` are translatable,
# we need to include them in the .po files.
# Gettext recognizes only strings given as parameters to the `_` function.
# To avoid initializing translations in this module we simply roll our own "fake" `_` function
# which returns whatever is given to it as an argument.
def _(arg):
	return arg


# Add-on information variables
addon_info = {
	# add-on Name/identifier, internal for NVDA
	"addon_name": "zPortapapeles",
	# Add-on summary, usually the user visible name of the addon.
	# Translators: Summary for this add-on
	# to be shown on installation and add-on information found in Add-ons Manager.
	"addon_summary": _("zPortapapeles"),
	# Add-on description
	# Translators: Long description to be shown for this add-on on add-on information from add-ons manager
	"addon_description": _("Utilidades para el portapapeles, anuncio de las teclas e historial."),
	# version
	"addon_version": "0.5.1",
	# Author(s)
	"addon_author": u"Héctor J. Benítez Corredera <xebolax@gmail.com> idea original Peter Vagner <peter.v@datagate.sk>",
	# URL for the add-on documentation support
	"addon_url": "https://github.com/hxebolax/zPortapapeles",
	# Documentation file name
	"addon_docFileName": "readme.html",
	# Minimum NVDA version supported (e.g. "2018.3.0", minor version is optional)
	"addon_minimumNVDAVersion": "2021.2.0",
	# Last NVDA version supported/tested (e.g. "2018.4.0", ideally more recent than minimum version)
	"addon_lastTestedNVDAVersion": "2023.1.0",
	# Add-on update channel (default is None, denoting stable releases,
	# and for development releases, use "dev".)
	# Do not change unless you know what you are doing!
	"addon_updateChannel": None,
}

import os.path

# Define the python files that are the sources of your add-on.
# You can either list every file (using ""/") as a path separator,
# or use glob expressions.
# For example to include all files with a ".py" extension from the "globalPlugins" dir of your add-on
# the list can be written as follows:
# pythonSources = ["addon/globalPlugins/*.py"]
# For more information on SCons Glob expressions please take a look at:
# https://scons.org/doc/production/HTML/scons-user/apd.html
pythonSources = [os.path.join("addon", "globalPlugins", "zPortapapeles", "*.py")]

# Files that contain strings for translation. Usually your python sources
i18nSources = ["buildVars.py"] + pythonSources

# Files that will be ignored when building the nvda-addon file
# Paths are relative to the addon directory, not to the root directory of your addon sources.
excludedFiles = []

# Base language for the NVDA add-on
# If the source code, add-on interface and readme file located in the root folder are written in a language other than english, modify this variable as appropriate.
# For example, set baseLanguage to "es" if your add-on is primarily in spanish
baseLanguage = "es"


addon/globalPlugins/zPortapapeles/portapapeles.py
# -*- coding: utf-8 -*-
# Copyright (C) 2022 Héctor J. Benítez Corredera <xebolax@gmail.com>
# This file is covered by the GNU General Public License.

import api
import ui
import wx
import nvwave
from tones import beep
from threading import Thread, Lock
from time import sleep
import os
import sys
import ctypes
from ctypes.wintypes import BOOL, HWND, HANDLE, HGLOBAL, UINT, LPVOID
from ctypes import c_size_t as SIZE_T
from typing import (Any, Callable, Dict, List, Iterable, Tuple) 
dirAddonPath=os.path.dirname(__file__)
sys.path.append(dirAddonPath)
import trans_clipboard as Trans
del sys.path[-1]
from . import ajustes

class ClipboardFormat:
    CF_TEXT = 1
    CF_BITMAP = 2
    CF_METAFILEPICT = 3
    CF_SYLK = 4
    CF_DIF = 5
    CF_TIFF = 6
    CF_OEMTEXT = 7
    CF_DIB = 8
    CF_PALETTE = 9
    CF_PENDATA = 10
    CF_RIFF = 11
    CF_WAVE = 12
    CF_UNICODETEXT = 13
    CF_ENHMETAFILE = 14
    CF_HDROP = 15
    CF_LOCALE = 16
    CF_DIBV5 = 17
    CF_MAX = 18
    CF_HTML = ctypes.windll.user32.RegisterClipboardFormatW("HTML Format")

# Solución extraida de https://stackoverflow.com/questions/579687/how-do-i-copy-a-string-to-the-clipboard
# Me gusta más esta forma que la interna de NVDA o la de wxpython para manejar el portapapeles, es mas directa con el sistema.

OpenClipboard = ctypes.windll.user32.OpenClipboard
OpenClipboard.argtypes = HWND,
OpenClipboard.restype = BOOL
EmptyClipboard = ctypes.windll.user32.EmptyClipboard
EmptyClipboard.restype = BOOL
GetClipboardData = ctypes.windll.user32.GetClipboardData
GetClipboardData.argtypes = UINT,
GetClipboardData.restype = HANDLE
SetClipboardData = ctypes.windll.user32.SetClipboardData
SetClipboardData.argtypes = UINT, HANDLE
SetClipboardData.restype = HANDLE
CloseClipboard = ctypes.windll.user32.CloseClipboard
CloseClipboard.restype = BOOL
CF_UNICODETEXT = 13

GlobalAlloc = ctypes.windll.kernel32.GlobalAlloc
GlobalAlloc.argtypes = UINT, SIZE_T
GlobalAlloc.restype = HGLOBAL
GlobalLock = ctypes.windll.kernel32.GlobalLock
GlobalLock.argtypes = HGLOBAL,
GlobalLock.restype = LPVOID
GlobalUnlock = ctypes.windll.kernel32.GlobalUnlock
GlobalUnlock.argtypes = HGLOBAL,
GlobalSize = ctypes.windll.kernel32.GlobalSize
GlobalSize.argtypes = HGLOBAL,
GlobalSize.restype = SIZE_T

GMEM_MOVEABLE = 0x0002
GMEM_ZEROINIT = 0x0040

unicode_type = type(u'')

def clean():
	# Función para borrar el portapapeles
	OpenClipboard(None)
	EmptyClipboard()
	CloseClipboard()

def get():
	# Función para obtener el contenido del portapapeles
	text = None
	OpenClipboard(None)
	handle = GetClipboardData(CF_UNICODETEXT)
	pcontents = GlobalLock(handle)
	size = GlobalSize(handle)
	if pcontents and size:
		raw_data = ctypes.create_string_buffer(size)
		ctypes.memmove(raw_data, pcontents, size)
		text = raw_data.raw.decode('utf-16le').rstrip(u'\0')
	GlobalUnlock(handle)
	CloseClipboard()
	return text

def put(text):
	# Función para pegar contenido al portapapeles
	if not isinstance(text, unicode_type):
		text = text.decode('mbcs')
	data = text.encode('utf-16le')
	OpenClipboard(None)
	EmptyClipboard()
	handle = GlobalAlloc(GMEM_MOVEABLE | GMEM_ZEROINIT, len(data) + 2)
	pcontents = GlobalLock(handle)
	ctypes.memmove(pcontents, data, len(data))
	GlobalUnlock(handle)
	SetClipboardData(CF_UNICODETEXT, handle)
	CloseClipboard()

def GetDictKeyName(theDict: Dict, theValue: Any, start: str = None) -> str:
	for key, value in theDict.items():
		if theValue == value and ((start and key.startswith(start)) or True):
			return key
	return ''

def detect() -> Dict[int, str]:
	formats = {}
	if OpenClipboard(None):
		formatType = 0
		arrayType = ctypes.c_wchar * 64
		while True:
			formatType = ctypes.windll.user32.EnumClipboardFormats(formatType)
			if formatType == 0:
				break
			values = arrayType()
			ctypes.windll.user32.GetClipboardFormatNameW(formatType, values, len(values))
			formatName = values.value
			if not formatName:
				formatName = GetDictKeyName(ClipboardFormat.__dict__, formatType, 'CF_')
			formats[formatType] = formatName
		CloseClipboard()
	if len(formats) == 0:
		return False
	else:
#		return formats
		if 'CF_UNICODETEXT' in formats.values():
			return True
		else:
			return False

def talk(text):
	# Función para mandar mensaje de lo copiado fuera del hilo de tiempo
#	traducir = traductor.translate(datos['description'], ajustes.langDict.get(ajustes.tempLang))
	ui.message(text)

def talkLang(text):
	# Función para mandar mensaje de lo copiado fuera del hilo de tiempo
	try:
		traducir = Trans.translate(text, ajustes.langDict.get(ajustes.langTrans))
		wx.CallAfter(ui.message, traducir)
	except:
		pass

class ClipMonitor(Thread):
# Clase de hilo con tiempo para monitorizar el contenido del portapapeles y actuar en consecuencia
	def __init__(self, tiempo, *args, **kwargs):
		Thread.__init__(self, *args, **kwargs)

		self.tiempo = tiempo
		self.__borrado = ""
		self.__historial = []
		self.__flag = True
		self.pausa = False
		self.__candado = Lock()
		self.fichero = os.path.join(os.path.dirname(__file__), "Sonidos", "historial.wav")
		self.banderaHabla = False
		self.ultimoClipboard = ""
		self.clipTemp = ""

	def run(self):
		while self.__flag:
			try:
				clip = get()
				total = len(self.__historial)
				self.__candado.acquire() # Cerramos el candado para evitar que un hilo intente leer al mismo tiempo que otro está escribiendo
				if self.pausa == False: # Si no hay pausa continuamos
					if clip is not None: # Si no hay nada en el portapapeles continuamos
						if self.exist(self.__historial, clip) == False: # Si no se encuentra en la lista
							if total == 0: # Comprobamos si la lista tiene entradas == 0
								if clip == self.__borrado: # Comprobamos si es igual que lo ultimo borrado
									pass
								else: # Si no es igual que lo ultimo borrado
									if clip == self.clipTemp:
										pass
									else:
										self.__borrado = ""
										self.__historial.insert(0, clip)
										if ajustes.talkPaste:
											self.banderaHabla = True
											self.ultimoClipboard = clip
											wx.CallAfter(self.onTalk)
										if ajustes.sndHistorial: # Si tenemos True sonido
											if os.path.exists(self.fichero):
												nvwave.playWaveFile(self.fichero)
											else:
												beep(1000,50)
							else: # La lista si tiene entradas
								if self.__historial[0] == clip: # Comprobamos si la ultima entrada es igual que portapapeles
									pass
								else: # No es igual que la ultima entrada
									if clip ==self.__borrado:
										pass
									else:
										if clip == self.clipTemp:
											pass
										else:
											self.__borrado = ""
											self.__historial.insert(0, clip)
											if ajustes.sndHistorial: # Si tenemos True sonido
												if os.path.exists(self.fichero):
													nvwave.playWaveFile(self.fichero)
												else:
													beep(1000,50)
						else:
							if not self.clipTemp == clip:
								self.clipTemp = clip
								if not self.banderaHabla:
									if ajustes.talkPaste:
										self.banderaHabla = True
										self.ultimoClipboard = clip
										wx.CallAfter(self.onTalk)

				self.__candado.release() # Abrimos el candado
				sleep(self.tiempo) # Hacemos una pausa de un segundo. Puede ser más o menos pero una mínima pausa hay que hacer si no probablemente se bloquearía.
			except OSError: # Se ha copiado un objeto que no es texto
				pass

	def onTalk(self):
		if self.banderaHabla:
			talk(self.ultimoClipboard)
			self.banderaHabla = False

	def exist(self, listado, buscar):
		if not buscar in listado:
			return False
		return True

	@property # Esto es un getter. Sirve para acceder a la variable __historial que es interna y a la que no se puede acceder desde fuera del objeto. Se invoca con clipMonitor.historial
	def historial(self):
		self.__candado.acquire()
		h = self.__historial
		self.__candado.release()
		return h

	def borrar(self, valor):
		self.__candado.acquire()
		self.__borrado = self.__historial[valor]
		del self.__historial[valor]
		self.__candado.release()

	def borrarTodo(self):
		self.__candado.acquire()
		self.__borrado = self.__historial[0]
		del self.__historial[:]
		self.__candado.release()

	def pausar(self):
		self.__candado.acquire()
		self.pausa = True

	def reanudar(self):
		self.pausa = False
		self.__candado.release()

	def SetTimer(self, valor):
		self.__candado.acquire()
		self.tiempo = valor
		self.__candado.release()

	def kill(self):
		self.__flag = False

class ClipMonitorGame(Thread):
# Clase de hilo con tiempo para monitorizar el contenido del portapapeles y actuar en consecuencia
	def __init__(self, tiempo, *args, **kwargs):
		Thread.__init__(self, *args, **kwargs)

		self.tiempo = tiempo
		self.__flag = True
		self.pausa = False
		self.__candado = Lock()
		self.temp = ""
		self.temp1 = ""
	def run(self):
		while self.__flag:
			try:
				clip = get()
				self.__candado.acquire() # Cerramos el candado para evitar que un hilo intente leer al mismo tiempo que otro está escribiendo
				self.temp = clip
				if self.pausa == False: # Si no hay pausa continuamos
					if clip is not None: # Si no hay nada en el portapapeles continuamos
						if self.temp == self.temp1:
							pass
						else:
							wx.CallAfter(self.onTalk, self.temp)
							self.temp1 = clip
				self.__candado.release() # Abrimos el candado
				sleep(self.tiempo) # Hacemos una pausa de un segundo. Puede ser más o menos pero una mínima pausa hay que hacer si no probablemente se bloquearía.
			except OSError: # Se ha copiado un objeto que no es texto
				pass

	def onTalk(self, texto):
		Thread(target=talkLang, args=[texto], daemon= True).start()
#		wx.CallAfter(talkLang, texto)

	def pausar(self):
		self.__candado.acquire()
		self.pausa = True

	def reanudar(self):
		self.pausa = False
		self.__candado.release()

	def SetTimer(self, valor):
		self.__candado.acquire()
		self.tiempo = valor
		self.__candado.release()

	def kill(self):
		self.__flag = False


addon/globalPlugins/zPortapapeles/ajustes.py
# -*- coding: utf-8 -*-
# Copyright (C) 2022 Héctor J. Benítez Corredera <xebolax@gmail.com>
# This file is covered by the GNU General Public License.

import addonHandler
import config

addonHandler.initTranslation()

messagesDict = {
	# Translators: Mensaje que informa de la acción de la combinación de tecla pulsada por el usuario
	"control+e":_("Seleccionar todo"),
	# Translators: Mensaje que informa de la acción de la combinación de tecla pulsada por el usuario
	"control+a":_("Seleccionar todo"),
	# Translators: Mensaje que informa de la acción de la combinación de tecla pulsada por el usuario
	"control+x":_("Cortar"),
	# Translators: Mensaje que informa de la acción de la combinación de tecla pulsada por el usuario
	"control+c":_("Copiar"),
	# Translators: Mensaje que informa de la acción de la combinación de tecla pulsada por el usuario
	"control+v":_("Pegar"),
	# Translators: Mensaje que informa de la acción de la combinación de tecla pulsada por el usuario
	"control+y":_("Rehacer"),
	# Translators: Mensaje que informa de la acción de la combinación de tecla pulsada por el usuario
	"control+z":_("Deshacer")
}

soundsDict = {
	_("Copiar"):"copiar.wav",
	_("Cortar"):"cortar.wav",
	_("Pegar"):"pegar.wav",
	_("Seleccionar todo"):"todo.wav",
	_("Rehacer"):"rehacer.wav",
	_("Deshacer"):"desacer.wav",
}

tiempoDict = {
	0:0.1,
	1:0.2,
	2:0.3,
	3:0.4,
	4:0.5,
	5:1,
}

tiempoChk = [
	# Translators: Descripción de item para el choice del tiempo de opciones
	_("1 décima de segundo"),
	# Translators: Descripción de item para el choice del tiempo de opciones
	_("2 décimas de segundo"),
	# Translators: Descripción de item para el choice del tiempo de opciones
	_("3 décimas de segundo"),
	# Translators: Descripción de item para el choice del tiempo de opciones
	_("4 décimas de segundo"),
	# Translators: Descripción de item para el choice del tiempo de opciones
	_("5 décimas de segundo"),
	# Translators: Descripción de item para el choice del tiempo de opciones
	_("1 segundo")
]

# Lista con idiomas para las traducciones
langLST = [_("Alemán"), _("Árabe"), _("Croata"), _("Español"), _("Francés"), _("Inglés"), _("Italiano"), _("Polaco"), _("Portugués"), _("Ruso"), _("Turco"), _("Ucraniano")]
# Diccionario con las abreviaturas de idioma
langDict = {
	0:"de",
	1:"ar",
	2:"hr",
	3:"es",
	4:"fr",
	5:"en",
	6:"it",
	7:"pl",
	8:"pt",
	9:"ru",
	10:"tr",
	11:"uk",
}

_main = None
audio = None
voz = None
historial = None
tiempo = None
sndHistorial = None
talkPaste = None
isGame = None
isActivo = None
tiempoLang = None
langTrans = None
winOn = False

def initConfiguration():
	confspec = {
		"voz": "boolean(default=True)",
		"audio": "boolean(default=False)",
		"historial": "boolean(default=False)",
		"tiempo": "integer(default=1, min=0, max=5)",
		"sonidoHistorial": "boolean(default=False)",
		"vozCopiado": "boolean(default=False)",
		"activado": "boolean(default=True)",
		"game": "boolean(default=False)",
		"tiempoLang": "integer(default=1, min=0, max=5)",
		"langTrans": "integer(default=5, min=0, max=11)",
	}
	config.conf.spec['zPortapapeles'] = confspec

def getConfig(key):
	value = config.conf["zPortapapeles"][key]
	return value

def setConfig(key, value):
	try:
		config.conf.profiles[0]["zPortapapeles"][key] = value
	except:
		config.conf["zPortapapeles"][key] = value

def setup():
	global voz, audio, historial, tiempo, sndHistorial, talkPaste, isGame, isActivo, tiempoLang, langTrans
	initConfiguration()
	voz = getConfig("voz")
	audio = getConfig("audio")
	historial = getConfig("historial")
	tiempo = getConfig("tiempo")
	sndHistorial = getConfig("sonidoHistorial")
	talkPaste = getConfig("vozCopiado")
	isGame = getConfig("game")
	isActivo = getConfig("activado")
	tiempoLang = getConfig("tiempoLang")
	langTrans = getConfig("langTrans")
